### PR TITLE
[FEA] Switch to `regex_program` APIs

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
Closes https://github.com/NVIDIA/spark-rapids/issues/7295.

This is the Spark plugin side work for switching to `regex_program` APIs, moving all existing java API calls to the version that takes in a `regex_program`.

Currently we documented that Java's split API produces different behaviors than cudf when splitting with empty pattern (in expressions `StringSplit` and `StringToMap`). We filed a follow up issue https://github.com/NVIDIA/spark-rapids/issues/7785 to handle this in the future.

### CODE REVIEW

**Comment by cindyyuanjiang** [comment]
build

**Review by revans2 — DISMISSED** [review]

**Comment by cindyyuanjiang** [comment]
We removed the check for empty string pattern for `stringSplit` and `stringSplitRecord` APIs in cudf (Java's split API produces different behaviors than cudf when splitting with empty pattern). Do we want to document this in the current PR? https://github.com/rapidsai/cudf/pull/12548#discussion_r1085453735

**Comment by cindyyuanjiang** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -2358,6 +2358,7 @@ object GpuOverrides extends Logging {
       }),
     expr[StringSplit](
        "Splits `str` around occurrences that match `regex`",
+      // Java's split API produces different behaviors than cudf when splitting with empty pattern
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2361]: So how is this being handled in Plugin ? Don't we need any change to take care of this case ?

Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2361]: We haven't discussed how to handle this in the Plugin. I will follow up with this point. Thank you!

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2361]: Better to file an issue to track this follow-up.

Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2361]: @firestarman Thank you! I filed issue: https://github.com/NVIDIA/spark-rapids/issues/7785 to track this (also updated in description above).

**Review by firestarman — APPROVED** [review]

**Pull request merged**